src/ioctls/device.rs

```diff
@@ -45,32 +47,71 @@ impl DeviceFd {
     /// # extern crate kvm_ioctls;
     /// # extern crate kvm_bindings;
     /// # use kvm_ioctls::Kvm;
+    /// # #[cfg(not(target_arch = "riscv64"))]
     /// # use kvm_bindings::{
-    ///    kvm_device_type_KVM_DEV_TYPE_VFIO,
-    ///    KVM_DEV_VFIO_GROUP, KVM_DEV_VFIO_GROUP_ADD, KVM_CREATE_DEVICE_TEST
-    /// };
+    /// #    kvm_device_type_KVM_DEV_TYPE_VFIO,
+    /// #    KVM_DEV_VFIO_GROUP, KVM_DEV_VFIO_GROUP_ADD, KVM_CREATE_DEVICE_TEST
+    /// # };
+    /// # #[cfg(target_arch = "riscv64")]
+    /// # use kvm_bindings::{
+    /// #     kvm_device_type_KVM_DEV_TYPE_RISCV_AIA,
+    /// #     KVM_DEV_RISCV_AIA_GRP_CONFIG, KVM_DEV_RISCV_AIA_CONFIG_SRCS
+    /// # };
     /// let kvm = Kvm::new().unwrap();
     /// let vm = kvm.create_vm().unwrap();
     ///
-    /// let mut device = kvm_bindings::kvm_create_device {
-    ///     type_: kvm_device_type_KVM_DEV_TYPE_VFIO,
-    ///     fd: 0,
-    ///     flags: KVM_CREATE_DEVICE_TEST,
-    /// };
-    ///
-    /// let device_fd = vm
-    ///     .create_device(&mut device)
-    ///     .expect("Cannot create KVM device");
-    ///
-    /// let dist_attr = kvm_bindings::kvm_device_attr {
-    ///     group: KVM_DEV_VFIO_GROUP,
-    ///     attr: u64::from(KVM_DEV_VFIO_GROUP_ADD),
-    ///     addr: 0x0,
-    ///     flags: 0,
-    /// };
-    ///
-    /// if (device_fd.has_device_attr(&dist_attr).is_ok()) {
-    ///     device_fd.set_device_attr(&dist_attr).unwrap();
+    /// // The VFIO device is not yet available on RISC-V. The patch for QEMU:
+    /// // https://lore.kernel.org/all/20240903201633.93182-1-dbarboza@ventanamicro.com/
+    /// // and patch for linux kernel
+    /// // https://github.com/ventanamicro/linux/tree/dev-upstream are both not
+    /// // upstreamed. Disabling VFIO device test fro RISC-V at the time being.
+    /// #[cfg(not(target_arch = "riscv64"))]
```

Comment:
Mh, is there another test we can instead do on riscv for this function? Otherwise, if its completely impossible to use this function on riscv yet, this should be part of the normal doccomment for this function, and not just the example. We should also use the `#` prefixes here so that the cfg and comment don't show up on docs.rs (and then have a little note at the top of the example that briefly says what it does, e.g. "seting up a vfio device on arm64")

Reply:
Good point, I'll fix it

Reply:
I moved the comment to normal document, and left cfg visible to tell the difference between riscv and no-riscv platform :)

Reply:
Well, then we could also add separate entries for all sorts of devices that are supported on each of the architectures. For such a low-level ioctl wrapper, I don't really think we should go down that route. E.g. using the functions in this crate, you're still very much expected to consult the KVM docs, and we probably shouldn't try to duplicate those here. The example for these imo should indicate what the general pattern of doing the ioctl should be (e.g. what structures to set up for the call, how to handle the return value), but I don't really see the value of doing that with multiple permutations of constants. And to prove that `set_device_attr` is functioning on riscv, we can add a unit tests (or if you think the AIA device is a better example than the VFIO device, then we keep the AIA example here, and move the VFIO example to the unittests).

That being said, maybe it's worth to add some slightly higher-level wrappers, say, `set_aia_device_attr` or something that specifically deals with AIA devices, of `set_vfio_device_attr` that specifically deals with VFIO devices? Then those should have examples in their doc comments detailing the specifics of their device.

Reply:
> That being said, maybe it's worth to add some slightly higher-level wrappers, say, `set_aia_device_attr` or something that specifically deals with AIA devices, or `set_vfio_device_attr` that specifically deals with VFIO devices? Then those should have examples in their doc comments detailing the specifics of their device.

Very good point, I'll work on it :)